packages/server/lib/browsers/chrome.ts

```diff
@@ -329,10 +328,100 @@ const _handleDownloads = async function (client, dir, automation) {
   })
 }
 
-const _setAutomation = (client, automation) => {
-  return automation.use(
-    new CdpAutomation(client.send, client.on, automation),
-  )
+let frameTree
+
+const _onBrowserPreRequest = (client) => {
+  return async (prerequest) => {
+  // this gets the frame tree ahead of the Fetch.requestPaused event, because
+  // the CDP is tied up during that event and can't be utilized. however,
+  // we avoid the overhead if it's not possible that the request will be for
+  // the AUT frame
+    if (
+      prerequest.originalResourceType !== 'Document'
+    || prerequest.url.includes('__cypress')
+    ) {
+      return
+    }
+
+    debug('get frame tree')
+
+    frameTree = (await client.send('Page.getFrameTree')).frameTree
```

Comment:
i don't think it's good to call `Page.getFrameTree` on every single prerequest. a decently sized cypress test suite can send thousands of requests, it seems wasteful to do this on each one. you could maintain a local cache of `frameTree`, initialized with `Page.getFrameTree`, then you could update it via calling `Page.getFrameTree` only when `Page.frameAttached/Detached/Navigated` are called, this would drastically cut down on CDP chatter. thoughts?

Reply:
Good idea. Addressed in https://github.com/cypress-io/cypress/pull/19323/commits/b3276ecc08cbe6fc04adce430833fd81b056760c